### Import Packages

In [464]:
import os

import glob

import pandas as pd

import zipfile
from pathlib import Path


import numpy as np
import pickle

### Notebook Settings

In [465]:
# Display Settings for Pandas DataFrames
pd.set_option("display.max_rows", None)         # How many rows to display when printing a DataFrame? None shows all rows.
pd.set_option("display.max_columns", None)      # How many columns to display when printing a DataFrame? None shows all columns.
pd.set_option("display.max_info_columns", 100)  # How many rows to display for .info()?
pd.set_option("display.max_info_rows", 1000000) # How many columns to display for .info()?
pd.set_option("display.precision", 2)           # How many decimals to show when printing a DataFrame?

### glob Function

In [466]:
# Find all .zip files recursively in data/external
zip_files = glob.glob('../../data/external/**/*.zip', recursive=True)

print(f"Found {len(zip_files)} zip file(s):")
for zip_file in zip_files:
    print(f"{zip_file}")

Found 27 zip file(s):
../../data/external/data-pack-2018-19/Knee Replacement CCG 1819.csv.zip
../../data/external/data-pack-2018-19/Hip Replacement CCG 1819.csv.zip
../../data/external/data-pack-2018-19/Hip Replacement Provider 1819.csv.zip
../../data/external/data-pack-2018-19/Time Series Hip and Knee Replacement 1819.csv.zip
../../data/external/data-pack-2018-19/ProvComm Hip and Knee Replacement 1819.csv.zip
../../data/external/data-pack-2018-19/PartLink Hip and Knee Replacement 1819.csv.zip
../../data/external/data-pack-2018-19/Key Facts Hip and Knee Replacements 1819.csv.zip
../../data/external/data-pack-2018-19/Knee Replacement Provider 1819.csv.zip
../../data/external/data-pack-2018-19/Equality 1819.csv.zip
../../data/external/data-pack-2017-18/Key Facts Hip and Knee Replacements 1718.csv.zip
../../data/external/data-pack-2017-18/Equality 1718.csv.zip
../../data/external/data-pack-2017-18/Hip Replacements CCG 1718.csv.zip
../../data/external/data-pack-2017-18/PartLink Hip and Kne

### Analyse First Selection

In [467]:
# 1. Define paths
# zip_path = "../../data/external/data-pack-2016-17/Knee Replacement Provider 1617.csv.zip"
# target_csv = "Knee Replacement Provider 1617.csv"


# 2. Open the zip and read the specific CSV
# with zipfile.ZipFile(zip_path, 'r') as z:
#    with z.open(target_csv) as f:
#        df_knee_provider_1617 = pd.read_csv(f)

# Display results
# print(f"Successfully loaded {target_csv}")
# print(df_knee_provider_1617.info())

### Load Files


In [468]:
#1. Define paths
zip_files = [
    "../../data/external/data-pack-2016-17/Knee Replacement Provider 1617.csv.zip",
    "../../data/external/data-pack-2017-18/Knee Replacements Provider 1718.csv.zip",
    "../../data/external/data-pack-2018-19/Knee Replacement Provider 1819.csv.zip",
]
# Select specific CSVs to read from each zip file
target_csv_files = [
    "Knee Replacement Provider 1617.csv",
    "Knee Replacements Provider 1718.csv",
    "Knee Replacement Provider 1819.csv"
]

dataframes = []

for zip_path, csv_name in zip(zip_files, target_csv_files):
    with zipfile.ZipFile(zip_path, 'r') as z:
        if csv_name in z.namelist():
            with z.open(csv_name) as f:
                df = pd.read_csv(f, low_memory=False)
                dataframes.append(df)
                print(f"Successfully loaded {csv_name} from {zip_path}")
        else:
            print(f"{csv_name} not found in {zip_path}")


Successfully loaded Knee Replacement Provider 1617.csv from ../../data/external/data-pack-2016-17/Knee Replacement Provider 1617.csv.zip
Successfully loaded Knee Replacements Provider 1718.csv from ../../data/external/data-pack-2017-18/Knee Replacements Provider 1718.csv.zip
Successfully loaded Knee Replacement Provider 1819.csv from ../../data/external/data-pack-2018-19/Knee Replacement Provider 1819.csv.zip


### Merge & Save File

In [469]:
# Merge all dataframes into a single dataframe
df_prov = pd.concat(dataframes, ignore_index=True)

# Save Output
df_prov.to_csv("df_knee_prov.csv", index=False)

print("Files Merged Successfully and saved as df_prov.csv")

Files Merged Successfully and saved as df_prov.csv


### Create a Copy

In [470]:
df_prov_copy = df_prov.copy(deep=False)

### Descriptive Statistics

Shape of the DataFrame

In [471]:
print("Shape of the merged dataframe:", df_prov.shape)

Shape of the merged dataframe: (139236, 81)


Show a sample of the first 3 rows

In [472]:
df_prov.head(3)

,Provider Code,Procedure,Revision Flag,Year,Age Band,Gender,Pre-Op Q Assisted,Pre-Op Q Assisted By,Pre-Op Q Symptom Period,Pre-Op Q Previous Surgery,Pre-Op Q Living Arrangements,Pre-Op Q Disability,Heart Disease,High Bp,Stroke,Circulation,Lung Disease,Diabetes,Kidney Disease,Nervous System,Liver Disease,Cancer,Depression,Arthritis,Pre-Op Q Mobility,Pre-Op Q Self-Care,Pre-Op Q Activity,Pre-Op Q Discomfort,Pre-Op Q Anxiety,Pre-Op Q EQ5D Index Profile,Pre-Op Q EQ5D Index,Post-Op Q Assisted,Post-Op Q Assisted By,Post-Op Q Living Arrangements,Post-Op Q Disability,Post-Op Q Mobility,Post-Op Q Self-Care,Post-Op Q Activity,Post-Op Q Discomfort,Post-Op Q Anxiety,Post-Op Q Satisfaction,Post-Op Q Sucess,Post-Op Q Allergy,Post-Op Q Bleeding,Post-Op Q Wound,Post-Op Q Urine,Post-Op Q Further Surgery,Post-Op Q Readmitted,Post-Op Q EQ5D Index Profile,Post-Op Q EQ5D Index,Knee Replacement EQ 5D Index Post-Op Q Predicted,Pre-Op Q EQ VAS,Post-Op Q EQ VAS,Knee Replacement EQ VAS_Post-Op Q Predicted,Knee Replacement Pre-Op Q Pain,Knee Replacement Pre-Op Q Night Pain,Knee Replacement Pre-Op Q Washing,Knee Replacement Pre-Op Q Transport,Knee Replacement Pre-Op Q Walking,Knee Replacement Pre-Op Q Standing,Knee Replacement Pre-Op Q Limping,Knee Replacement Pre-Op Q Kneeling,Knee Replacement Pre-Op Q Work,Knee Replacement Pre-Op Q Confidence,Knee Replacement Pre-Op Q Shopping,Knee Replacement Pre-Op Q Stairs,Knee Replacement Pre-Op Q Score,Knee Replacement Post-Op Q Pain,Knee Replacement Post-Op Q Night Pain,Knee Replacement Post-Op Q Washing,Knee Replacement Post-Op Q Transport,Knee Replacement Post-Op Q Walking,Knee Replacement Post-Op Q Standing,Knee Replacement Post-Op Q Limping,Knee Replacement Post-Op Q Kneeling,Knee Replacement Post-Op Q Work,Knee Replacement Post-Op Q Confidence,Knee Replacement Post-Op Q Shopping,Knee Replacement Post-Op Q Stairs,Knee Replacement Post-Op Q Score,Knee Replacement OKS Post-Op Q Predicted
0,ADP02,Knee Replacement,0,2016/17,*,*,2,0,4,2,2,1,9,9,9,9,9,9,9,9,9,9,1,9,2,2,2,3,2,22232,-0.02,2,9,2,1,2,9,2,2,3,2,1,2,2,2,2,2,2,29223,NaN,NaN,63,50,59.82,0,0,2,3,2,1,0,0,1,2,2,2,15.0,3,2,3,3,2,2,3,2,3,1,3,2,29.0,28.12
1,ADP02,Knee Replacement,0,2016/17,*,*,2,0,1,2,2,2,9,9,9,9,9,9,9,9,9,9,1,9,2,2,3,2,2,22322,0.19,2,9,2,2,1,1,2,2,2,3,1,1,2,2,2,2,1,11222,0.69,0.67,30,58,64.73,3,2,4,3,4,3,1,1,1,3,2,3,30.0,4,1,3,2,4,4,4,2,4,4,4,3,39.0,39.59
2,ADP02,Knee Replacement,0,2016/17,*,*,2,0,2,2,1,2,9,9,9,9,9,9,9,9,9,9,9,1,2,1,3,3,1,21331,0.10,2,9,1,2,1,1,1,1,1,2,1,2,2,2,2,2,2,11111,1.00,0.79,65,80,77.92,0,1,4,2,2,1,2,3,1,3,2,3,24.0,3,3,4,4,4,4,4,3,4,4,4,4,45.0,40.25


### Data Types

In [473]:
df_prov.dtypes

Provider Code                                        object
Procedure                                            object
Revision Flag                                         int64
Year                                                 object
Age Band                                             object
Gender                                               object
Pre-Op Q Assisted                                     int64
Pre-Op Q Assisted By                                  int64
Pre-Op Q Symptom Period                               int64
Pre-Op Q Previous Surgery                             int64
Pre-Op Q Living Arrangements                          int64
Pre-Op Q Disability                                   int64
Heart Disease                                         int64
High Bp                                               int64
Stroke                                                int64
Circulation                                           int64
Lung Disease                            

### Dataframe Info

In [474]:
df_prov.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 139236 entries, 0 to 139235
Columns: 81 entries, Provider Code to Knee Replacement OKS Post-Op Q Predicted
dtypes: float64(7), int64(69), object(5)
memory usage: 86.0+ MB


### Analyse Revision Surgery

In [475]:
# Count the number of primary (0) and revision (1) surgeries, sorted by flag value
df_prov["Revision Flag"].value_counts().sort_index()

Revision Flag
0    133962
1      5274
Name: count, dtype: int64

In [476]:
# Count patients by previous surgery status (1=yes, 2=no, 9=not recorded), sorted by value
df_prov["Pre-Op Q Previous Surgery"].value_counts().sort_index()

Pre-Op Q Previous Surgery
1     10784
2    127410
9      1042
Name: count, dtype: int64

### Split into Numerical & String DataFrames

In [477]:
# Create table with numerical data
df_prov_num = df_prov.select_dtypes(include=["number"])

# Column names
l_df_num_names = df_prov_num.columns.tolist()

print(f"We have {len(l_df_num_names)} numerical variables:\n{l_df_num_names}")

We have 76 numerical variables:
['Revision Flag', 'Pre-Op Q Assisted', 'Pre-Op Q Assisted By', 'Pre-Op Q Symptom Period', 'Pre-Op Q Previous Surgery', 'Pre-Op Q Living Arrangements', 'Pre-Op Q Disability', 'Heart Disease', 'High Bp', 'Stroke', 'Circulation', 'Lung Disease', 'Diabetes', 'Kidney Disease', 'Nervous System', 'Liver Disease', 'Cancer', 'Depression', 'Arthritis', 'Pre-Op Q Mobility', 'Pre-Op Q Self-Care', 'Pre-Op Q Activity', 'Pre-Op Q Discomfort', 'Pre-Op Q Anxiety', 'Pre-Op Q EQ5D Index Profile', 'Pre-Op Q EQ5D Index', 'Post-Op Q Assisted', 'Post-Op Q Assisted By', 'Post-Op Q Living Arrangements', 'Post-Op Q Disability', 'Post-Op Q Mobility', 'Post-Op Q Self-Care', 'Post-Op Q Activity', 'Post-Op Q Discomfort', 'Post-Op Q Anxiety', 'Post-Op Q Satisfaction', 'Post-Op Q Sucess', 'Post-Op Q Allergy', 'Post-Op Q Bleeding', 'Post-Op Q Wound', 'Post-Op Q Urine', 'Post-Op Q Further Surgery', 'Post-Op Q Readmitted', 'Post-Op Q EQ5D Index Profile', 'Post-Op Q EQ5D Index', 'Knee Repl

In [478]:
# Create table with string data
df_prov_str = df_prov.select_dtypes(include=["object"])

# Column names
l_df_str_names = df_prov_str.columns.tolist()

print(f"We have {len(l_df_str_names)} string variables:\n{l_df_str_names}")

We have 5 string variables:
['Provider Code', 'Procedure', 'Year', 'Age Band', 'Gender']


CHECK: Is Gender a String Variable?

### Cross-Check!!!

In [479]:
print(
    f"The original data has  {df_prov.shape[1]} columns. The number and string data" 
    f"have total of {df_prov_num.shape[1] + df_prov_str.shape[1]} columns."
)

The original data has  81 columns. The number and string datahave total of 81 columns.


In [480]:
print(f"The number of obeservations: {df_prov.shape[0]}")

The number of obeservations: 139236


In [481]:
# We use the number of observations more often, so we define a variable.
n_obs = df_prov.shape[0]

### Check Missing Variables

#### Numerical Check

In [482]:
df_prov_missing_data_num = (
    
    pd.DataFrame({

        'data_type':    df_prov_num.dtypes,
        'missing':      df_prov_num.isnull().sum()   
    })

    # Add complete%
    .assign(
        missing_pct = lambda x: round(100 * x['missing'] / n_obs, 2),
        complete_pct = lambda x: round(100 * (n_obs - x['missing']) / n_obs, 2)
    )

    # Remove variables that are complete.
    .query("complete_pct < 100")

    # Sort table by data completeness in ascending order (highest % incomplete first).
    .sort_values(by = 'complete_pct')
)

df_prov_missing_data_num


,data_type,missing,missing_pct,complete_pct
Knee Replacement EQ VAS_Post-Op Q Predicted,float64,19423,13.95,86.05
Knee Replacement EQ 5D Index Post-Op Q Predicted,float64,14322,10.29,89.71
Pre-Op Q EQ5D Index,float64,7500,5.39,94.61
Post-Op Q EQ5D Index,float64,6274,4.51,95.49
Knee Replacement OKS Post-Op Q Predicted,float64,5381,3.86,96.14
Knee Replacement Post-Op Q Score,float64,2579,1.85,98.15
Knee Replacement Pre-Op Q Score,float64,1669,1.20,98.80


#### String Check

In [483]:
# Pandas.
df_prov_missing_data_str = (
    
    pd.DataFrame({

        'data_type':    df_prov_str.dtypes,
        'missing':      df_prov_str.isnull().sum()   
    })

    # Add complete%
    .assign(
        complete_pct = lambda x: round(100 * (n_obs - x['missing']) / n_obs, 2)
    )

    # Remove variables that are complete.
    #.query("complete_pct < 100")

    # Sort table by number of missing data in ascending order.
    .sort_values(
        by        = 'complete_pct',
        ascending = False
    )
)

df_prov_missing_data_str

,data_type,missing,complete_pct
Provider Code,object,0,100.0
Procedure,object,0,100.0
Year,object,0,100.0
Age Band,object,0,100.0
Gender,object,0,100.0


### Descriptive Statistics

#### Numeric Statistics

In [484]:
df_prov_num.describe()

,Revision Flag,Pre-Op Q Assisted,Pre-Op Q Assisted By,Pre-Op Q Symptom Period,Pre-Op Q Previous Surgery,Pre-Op Q Living Arrangements,Pre-Op Q Disability,Heart Disease,High Bp,Stroke,Circulation,Lung Disease,Diabetes,Kidney Disease,Nervous System,Liver Disease,Cancer,Depression,Arthritis,Pre-Op Q Mobility,Pre-Op Q Self-Care,Pre-Op Q Activity,Pre-Op Q Discomfort,Pre-Op Q Anxiety,Pre-Op Q EQ5D Index Profile,Pre-Op Q EQ5D Index,Post-Op Q Assisted,Post-Op Q Assisted By,Post-Op Q Living Arrangements,Post-Op Q Disability,Post-Op Q Mobility,Post-Op Q Self-Care,Post-Op Q Activity,Post-Op Q Discomfort,Post-Op Q Anxiety,Post-Op Q Satisfaction,Post-Op Q Sucess,Post-Op Q Allergy,Post-Op Q Bleeding,Post-Op Q Wound,Post-Op Q Urine,Post-Op Q Further Surgery,Post-Op Q Readmitted,Post-Op Q EQ5D Index Profile,Post-Op Q EQ5D Index,Knee Replacement EQ 5D Index Post-Op Q Predicted,Pre-Op Q EQ VAS,Post-Op Q EQ VAS,Knee Replacement EQ VAS_Post-Op Q Predicted,Knee Replacement Pre-Op Q Pain,Knee Replacement Pre-Op Q Night Pain,Knee Replacement Pre-Op Q Washing,Knee Replacement Pre-Op Q Transport,Knee Replacement Pre-Op Q Walking,Knee Replacement Pre-Op Q Standing,Knee Replacement Pre-Op Q Limping,Knee Replacement Pre-Op Q Kneeling,Knee Replacement Pre-Op Q Work,Knee Replacement Pre-Op Q Confidence,Knee Replacement Pre-Op Q Shopping,Knee Replacement Pre-Op Q Stairs,Knee Replacement Pre-Op Q Score,Knee Replacement Post-Op Q Pain,Knee Replacement Post-Op Q Night Pain,Knee Replacement Post-Op Q Washing,Knee Replacement Post-Op Q Transport,Knee Replacement Post-Op Q Walking,Knee Replacement Post-Op Q Standing,Knee Replacement Post-Op Q Limping,Knee Replacement Post-Op Q Kneeling,Knee Replacement Post-Op Q Work,Knee Replacement Post-Op Q Confidence,Knee Replacement Post-Op Q Shopping,Knee Replacement Post-Op Q Stairs,Knee Replacement Post-Op Q Score,Knee Replacement OKS Post-Op Q Predicted
count,139236.00,139236.00,139236.0,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,131736.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,132962.00,124914.00,139236.00,139236.00,119813.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,137567.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,139236.00,136657.00,133855.00
mean,0.04,1.92,0.0,2.65,1.97,1.35,1.81,8.25,5.46,8.87,8.55,8.27,8.00,8.84,8.92,8.95,8.57,8.25,2.84,2.14,1.55,2.26,2.64,1.68,23214.76,0.41,2.01,8.41,1.41,1.93,1.61,1.32,1.70,1.87,1.42,2.36,1.53,2.53,2.80,2.64,2.68,2.06,2.02,17587.27,0.75,0.72,154.01,118.23,72.58,0.58,1.30,2.82,2.13,2.12,1.75,0.97,0.90,1.49,1.91,1.91,1.85,18.98,2.67,2.85,3.62,3.11,3.39,3.12,3.12,1.65,3.09,3.53,3.21,3.06,35.89,34.43
std,0.19,0.82,0.0,1.05,0.67,1.04,1.59,2.33,3.97,1.02,1.85,2.31,2.64,1.13,0.80,0.61,1.81,2.33,3.36,1.25,1.43,1.31,1.39,1.54,13893.51,0.31,0.79,2.05,1.20,1.47,1.20,1.15,1.23,1.34,1.25,1.33,1.23,2.08,2.34,2.21,2.26,0.81,0.85,12939.07,0.25,0.12,271.43,196.45,9.46,0.72,1.37,1.02,1.07,1.30,1.11,1.29,1.18,1.14,1.36,1.39,1.14,7.75,1.29,1.34,0.90,1.06,1.13,1.03,1.21,1.48,1.11,0.95,1.29,1.14,9.47,4.74
min,0.00,1.00,0.0,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,11111.00,-0.59,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,11111.00,-0.59,0.03,0.00,0.00,21.76,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,7.06
25%,0.00,2.00,0.0,2.00,2.00,1.00,1.00,9.00,1.00,9.00,9.00,9.00,9.00,9.00,9.00,9.00,9.00,9.00,1.00,2.00,1.00,2.00,2.00,1.00,21221.00,0.09,2.00,9.00,1.00,1.0

#### String Statistics

In [485]:
df_prov_str.describe()

,Provider Code,Procedure,Year,Age Band,Gender
count,139236,139236,139236,139236,139236
unique,294,1,3,7,3
top,RL1,Knee Replacement,2016/17,70 to 79,2
freq,3135,139236,47604,55056,74085


!!! EXTRA: Consider developing f_describe (utils package)

### Concatenate numerical & string data frames

In [486]:
# Concatenate numerical and string dataframes back into one dataframe
df_prov_imp = pd.concat([df_prov_str, df_prov_num], axis=1)

df_prov_imp.shape

(139236, 81)

Check: is table will concatenated? 

In [487]:
df_prov_imp.head(3)

,Provider Code,Procedure,Year,Age Band,Gender,Revision Flag,Pre-Op Q Assisted,Pre-Op Q Assisted By,Pre-Op Q Symptom Period,Pre-Op Q Previous Surgery,Pre-Op Q Living Arrangements,Pre-Op Q Disability,Heart Disease,High Bp,Stroke,Circulation,Lung Disease,Diabetes,Kidney Disease,Nervous System,Liver Disease,Cancer,Depression,Arthritis,Pre-Op Q Mobility,Pre-Op Q Self-Care,Pre-Op Q Activity,Pre-Op Q Discomfort,Pre-Op Q Anxiety,Pre-Op Q EQ5D Index Profile,Pre-Op Q EQ5D Index,Post-Op Q Assisted,Post-Op Q Assisted By,Post-Op Q Living Arrangements,Post-Op Q Disability,Post-Op Q Mobility,Post-Op Q Self-Care,Post-Op Q Activity,Post-Op Q Discomfort,Post-Op Q Anxiety,Post-Op Q Satisfaction,Post-Op Q Sucess,Post-Op Q Allergy,Post-Op Q Bleeding,Post-Op Q Wound,Post-Op Q Urine,Post-Op Q Further Surgery,Post-Op Q Readmitted,Post-Op Q EQ5D Index Profile,Post-Op Q EQ5D Index,Knee Replacement EQ 5D Index Post-Op Q Predicted,Pre-Op Q EQ VAS,Post-Op Q EQ VAS,Knee Replacement EQ VAS_Post-Op Q Predicted,Knee Replacement Pre-Op Q Pain,Knee Replacement Pre-Op Q Night Pain,Knee Replacement Pre-Op Q Washing,Knee Replacement Pre-Op Q Transport,Knee Replacement Pre-Op Q Walking,Knee Replacement Pre-Op Q Standing,Knee Replacement Pre-Op Q Limping,Knee Replacement Pre-Op Q Kneeling,Knee Replacement Pre-Op Q Work,Knee Replacement Pre-Op Q Confidence,Knee Replacement Pre-Op Q Shopping,Knee Replacement Pre-Op Q Stairs,Knee Replacement Pre-Op Q Score,Knee Replacement Post-Op Q Pain,Knee Replacement Post-Op Q Night Pain,Knee Replacement Post-Op Q Washing,Knee Replacement Post-Op Q Transport,Knee Replacement Post-Op Q Walking,Knee Replacement Post-Op Q Standing,Knee Replacement Post-Op Q Limping,Knee Replacement Post-Op Q Kneeling,Knee Replacement Post-Op Q Work,Knee Replacement Post-Op Q Confidence,Knee Replacement Post-Op Q Shopping,Knee Replacement Post-Op Q Stairs,Knee Replacement Post-Op Q Score,Knee Replacement OKS Post-Op Q Predicted
0,ADP02,Knee Replacement,2016/17,*,*,0,2,0,4,2,2,1,9,9,9,9,9,9,9,9,9,9,1,9,2,2,2,3,2,22232,-0.02,2,9,2,1,2,9,2,2,3,2,1,2,2,2,2,2,2,29223,NaN,NaN,63,50,59.82,0,0,2,3,2,1,0,0,1,2,2,2,15.0,3,2,3,3,2,2,3,2,3,1,3,2,29.0,28.12
1,ADP02,Knee Replacement,2016/17,*,*,0,2,0,1,2,2,2,9,9,9,9,9,9,9,9,9,9,1,9,2,2,3,2,2,22322,0.19,2,9,2,2,1,1,2,2,2,3,1,1,2,2,2,2,1,11222,0.69,0.67,30,58,64.73,3,2,4,3,4,3,1,1,1,3,2,3,30.0,4,1,3,2,4,4,4,2,4,4,4,3,39.0,39.59
2,ADP02,Knee Replacement,2016/17,*,*,0,2,0,2,2,1,2,9,9,9,9,9,9,9,9,9,9,9,1,2,1,3,3,1,21331,0.10,2,9,1,2,1,1,1,1,1,2,1,2,2,2,2,2,2,11111,1.00,0.79,65,80,77.92,0,1,4,2,2,1,2,3,1,3,2,3,24.0,3,3,4,4,4,4,4,3,4,4,4,4,45.0,40.25


### Feature Selection

This cell selects the relevant columns from the full dataset and stores them in `df_prov_imp`. The selected features cover four categories:

- **Demographics** — `Age Band`, `Gender`
- **Pre-op survey responses** — patient-reported symptom duration, prior surgery, living arrangements, and disability status
- **Comorbidities** — 10 binary flags indicating chronic conditions (e.g. heart disease, hypertension, diabetes, cancer)
- **Oxford Knee Score (OKS) — Pre-op items** — 11 functional/pain questions (e.g. walking, kneeling, stairs) plus the composite pre-op score

The **outcome variable** is health gain, defined as `Knee Replacement Post-Op Q Score` - `Knee Replacement Pre-Op Q Score`, which measures patient-reported health gain after surgery. All other columns in the original dataset are dropped.

In [488]:
keep_cols = [
    "Age Band",
    "Gender",
    "Pre-Op Q Symptom Period",
    "Pre-Op Q Previous Surgery",
    "Pre-Op Q Living Arrangements",
    "Pre-Op Q Disability",
    "Heart Disease",
    "High Bp",
    "Stroke",
    "Circulation",
    "Lung Disease",
    "Diabetes",
    "Kidney Disease",
    "Nervous System",
    "Liver Disease",
    "Cancer",
    "Depression",
    "Arthritis",
    "Knee Replacement Pre-Op Q Pain",
    "Knee Replacement Pre-Op Q Night Pain",
    "Knee Replacement Pre-Op Q Washing",
    "Knee Replacement Pre-Op Q Transport",
    "Knee Replacement Pre-Op Q Walking",
    "Knee Replacement Pre-Op Q Standing",
    "Knee Replacement Pre-Op Q Limping",
    "Knee Replacement Pre-Op Q Kneeling",
    "Knee Replacement Pre-Op Q Work",
    "Knee Replacement Pre-Op Q Confidence",
    "Knee Replacement Pre-Op Q Shopping",
    "Knee Replacement Pre-Op Q Stairs",
    "Knee Replacement Pre-Op Q Score",
    "Knee Replacement Post-Op Q Score",
]

df_prov_slc = df_prov_imp[keep_cols].copy()

In [489]:
df_prov_slc.shape

(139236, 32)

### Rescaling - Standardization missing-value codes → `NaN`

The NHS PROMs data uses several conventions for missing / suppressed values:
- **`9`** — the dictionary-defined missing code for most categorical fields
- **`*`** — statistical suppression applied to small counts (< 5) by NHS Digital
- **`999`** — occasionally used for out-of-range / missing numeric entries
- **`''`** — empty string from CSV export

Converting all of these to `NaN` before any imputation or deletion step ensures that
downstream pandas methods (`.isna()`, `.fillna()`, `.dropna()`) behave consistently.

In [490]:
#df_prov_slc.replace({9: np.nan, "*": np.nan, "": np.nan, 999: np.nan}, inplace=True)

In [491]:
#total_missing = df_prov_slc.isna().sum().sum()
#print(f'Total missing cells after standardisation: {total_missing:,}')

### Missingness

This cell audits missing data across the selected feature columns in `df_prov_imp`.

For each column it computes:
- **`missing`** — count of `NaN` values *or* NHS suppression markers (`*`), which were not caught by the earlier standardisation step
- **`complete_pct`** — percentage of rows that are non-missing

The resulting summary table `df_prov_imp_missing` is sorted ascending by completeness so the most problematic columns appear first.

This informs downstream decisions about imputation strategy: columns with low completeness may need a different approach than near-complete ones.

In [492]:
# Pandas.
df_prov_slc_missing = (
    
    pd.DataFrame({

        'data_type':    df_prov_slc.dtypes,
        'missing':      (df_prov_slc.isnull() | df_prov_slc.eq('*')).sum()   
    })

    # Add complete%
    .assign(
        complete_pct = lambda x: round(100 * (n_obs - x['missing']) / n_obs, 2)
    )

    # Remove variables that are complete.
    #.query("complete_pct < 100")

    # Sort table by number of missing data in ascending order.
    .sort_values(
        by        = 'complete_pct',
        ascending = True
    )
)

df_prov_slc_missing

,data_type,missing,complete_pct
Age Band,object,9402,93.25
Gender,object,9402,93.25
Knee Replacement Post-Op Q Score,float64,2579,98.15
Knee Replacement Pre-Op Q Score,float64,1669,98.80
Pre-Op Q Living Arrangements,int64,0,100.00
Pre-Op Q Disability,int64,0,100.00
Pre-Op Q Symptom Period,int64,0,100.00
Pre-Op Q Previous Surgery,int64,0,100.00
Stroke,int64,0,100.00
Circulation,int64,0,100.00


### Scoping (Age, Gender, etc)

Remove Unknown Age Bands

In [493]:
dem_cols = [
    "Age Band",
    "Gender"
]

# Replace "*" with NaN to treat as missing values
df_prov_slc[dem_cols] = df_prov_slc[dem_cols].replace("*", None)

# Drop rows with missing age and gender
df_prov_slc = df_prov_slc.dropna(subset=dem_cols)
print(f"Shape of the dataframe after dropping rows with missing age and gender: {df_prov_slc.shape}")

Shape of the dataframe after dropping rows with missing age and gender: (129834, 32)


In [494]:
# Count patients by Age, sorted by value
print(df_prov_slc["Age Band"].value_counts().sort_index())

# Count all patients
df_prov_slc["Age Band"].value_counts().sum()

Age Band
40 to 49       250
50 to 59     13260
60 to 69     45361
70 to 79     55056
80 to 89     15883
90 to 120       24
Name: count, dtype: int64


np.int64(129834)

In [495]:
# Count patients by Age, sorted by value
print(df_prov_slc["Gender"].value_counts().sort_index())

# Count all patients
df_prov_slc["Gender"].value_counts().sum()

Gender
1    55749
2    74085
Name: count, dtype: int64


np.int64(129834)

Delete Age Bands 90 to 120?

### Recode: Comorbidity Flags

These 12 columns are binary indicators (1 = condition present, 9 = not reported).  
In line with standard epidemiological practice for administrative datasets, a missing
comorbidity flag is treated as *not present* (0) rather than excluded.  
Rationale: patients with unrecorded comorbidities are unlikely to differ systematically
from patients who answered 'No'; dropping these rows would introduce selection bias.

In [496]:
comorb_cols = [
    "Arthritis", "Cancer", "Circulation", "Depression", "Diabetes",
    "Heart Disease", "High Bp", "Kidney Disease", "Liver Disease",
    "Lung Disease", "Nervous System", "Stroke"
]

# For comorbidities, 9 = not recorded → treat as 0 (no comorbidity)
for col in comorb_cols:
    df_prov_slc[col] = df_prov_slc[col].replace(9, 0).astype(int)

print(df_prov_slc[comorb_cols].mean().sort_values().round(2)) 

print(
    f"The imputed data has {df_prov_slc['Arthritis'].isnull().sum().sum()} "
    f"missing values."
)


Liver Disease     0.01
Nervous System    0.01
Stroke            0.02
Kidney Disease    0.02
Cancer            0.05
Circulation       0.06
Depression        0.09
Lung Disease      0.09
Heart Disease     0.10
Diabetes          0.13
High Bp           0.45
Arthritis         0.77
dtype: float64
The imputed data has 0 missing values.


In [500]:
# Count patients by Comorbidities, sorted by value
for col in ["Arthritis", 
            "Cancer", 
            "Circulation", 
            "Depression", 
            "Diabetes",
            "Heart Disease", 
            "High Bp", 
            "Kidney Disease", 
            "Liver Disease",
            "Lung Disease", 
            "Nervous System", 
            "Stroke"]:
    print(f"\n{col}:")
    print(df_prov_slc[col].value_counts().sort_index())


Arthritis:
Arthritis
0     29552
1    100282
Name: count, dtype: int64

Cancer:
Cancer
0    122694
1      7140
Name: count, dtype: int64

Circulation:
Circulation
0    122396
1      7438
Name: count, dtype: int64

Depression:
Depression
0    117851
1     11983
Name: count, dtype: int64

Diabetes:
Diabetes
0    113334
1     16500
Name: count, dtype: int64

Heart Disease:
Heart Disease
0    117387
1     12447
Name: count, dtype: int64

High Bp:
High Bp
0    71626
1    58208
Name: count, dtype: int64

Kidney Disease:
Kidney Disease
0    127137
1      2697
Name: count, dtype: int64

Liver Disease:
Liver Disease
0    129068
1       766
Name: count, dtype: int64

Lung Disease:
Lung Disease
0    117714
1     12120
Name: count, dtype: int64

Nervous System:
Nervous System
0    128480
1      1354
Name: count, dtype: int64

Stroke:
Stroke
0    127645
1      2189
Name: count, dtype: int64


Verify that all Comorbidity is complete

In [499]:
# Verify that all 9s have been replaced
comorb_cols = [
    'Arthritis', 'Cancer', 'Circulation', 'Depression', 'Diabetes',
    'Heart Disease', 'High Bp', 'Kidney Disease', 'Liver Disease',
    'Lung Disease', 'Nervous System', 'Stroke'
]

# Check which indicator columns exist and have null values
existing_cols = [c for c in comorb_cols if c in df_prov_slc.columns]
missing_cols  = [c for c in comorb_cols if c not in df_prov_slc.columns]

if missing_cols:
    print(f"Columns NOT found in dataframe: {missing_cols}")

null_check = df_prov_slc[existing_cols].isnull().sum()
print("Null counts for indicator columns:")
print(null_check)


Null counts for indicator columns:
Arthritis         0
Cancer            0
Circulation       0
Depression        0
Diabetes          0
Heart Disease     0
High Bp           0
Kidney Disease    0
Liver Disease     0
Lung Disease      0
Nervous System    0
Stroke            0
dtype: int64


### Remove all Missing Pre-Op and Post-Op Missing Scores

In [501]:
# Count patients by OKS Pre-Op Dimensions, sorted by value
for col in ["Knee Replacement Pre-Op Q Pain",
    "Knee Replacement Pre-Op Q Night Pain",
    "Knee Replacement Pre-Op Q Washing",
    "Knee Replacement Pre-Op Q Transport",
    "Knee Replacement Pre-Op Q Walking",
    "Knee Replacement Pre-Op Q Standing",
    "Knee Replacement Pre-Op Q Limping",
    "Knee Replacement Pre-Op Q Kneeling",
    "Knee Replacement Pre-Op Q Work",
    "Knee Replacement Pre-Op Q Confidence",
    "Knee Replacement Pre-Op Q Shopping",
    "Knee Replacement Pre-Op Q Stairs",
    "Knee Replacement Pre-Op Q Score",
    "Knee Replacement Post-Op Q Score"]:
    print(f"\n{col}:")
    print(df_prov_slc[col].value_counts().sort_index())


Knee Replacement Pre-Op Q Pain:
Knee Replacement Pre-Op Q Pain
0    65650
1    56745
2     5731
3     1229
4      298
9      181
Name: count, dtype: int64

Knee Replacement Pre-Op Q Night Pain:
Knee Replacement Pre-Op Q Night Pain
0    41840
1    38259
2    33570
3     6875
4     8064
9     1226
Name: count, dtype: int64

Knee Replacement Pre-Op Q Washing:
Knee Replacement Pre-Op Q Washing
0      735
1    11274
2    40735
3    35590
4    41386
9      114
Name: count, dtype: int64

Knee Replacement Pre-Op Q Transport:
Knee Replacement Pre-Op Q Transport
0      684
1    29739
2    67384
3    22286
4     8466
9     1275
Name: count, dtype: int64

Knee Replacement Pre-Op Q Walking:
Knee Replacement Pre-Op Q Walking
0    15755
1    16256
2    52724
3    33448
4    10291
9     1360
Name: count, dtype: int64

Knee Replacement Pre-Op Q Standing:
Knee Replacement Pre-Op Q Standing
0     4399
1    55950
2    47471
3    18291
4     2354
9     1369
Name: count, dtype: int64

Knee Replacement Pre-

### Compare Imputed vs Non-Imputed DataFrames

In [ ]:
# Compare Results
pd.DataFrame({
    "original": df_prov_slc['Arthritis'].describe(),
    "imputed": df_prov_slc['Arthritis'].describe()
})

,original,imputed
count,129834.00,129834.00
mean,0.77,0.77
std,0.42,0.42
min,0.00,0.00
25%,1.00,1.00
50%,1.00,1.00
75%,1.00,1.00
max,1.00,1.00
